### PR TITLE
SONARIAC-1614 Improve precision of S1874 to reduce the FP rate

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: sonar-helm-for-iac/go.sum
Comment (petertrr) [inline sonar-helm-for-iac/go.sum:1]: This is the result of executing `go mod tidy`

File: iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{{/*
+Return the appropriate apiVersion for deployment.
+*/}}
+{{- define "deployment.apiVersion" -}}
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/_helpers.tpl:4]: The name `deployment.apiVersion` is not the best as `apiVersion` is used in many places and may confuse us. Maybe `deployment.gitVersion` will be better?

File: iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/deprecated_code.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+spec:
+  # Noncompliant@+1
+  {{- if semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion }}
+  ingressClassName: {{ .Values.ingress.className |  default "nginx" }}
+  {{- end }}
+  # Noncompliant@+1
+  {{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion)) }}
+  ingressClassName: "test"
+  {{- end }}
+  # It's not possible to raise inside the usage of "Capabilities.KubeVersion.GitVersion" inside the .tpl file
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/deprecated_code.yaml:29]: Do we have a ticket to address it? If not please create a new ticket.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmParser.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ private FileTree evaluateAndParseHelmFile(String source, HelmInputFileContext in
   }
 
   static boolean isInvalidHelmInputFile(HelmInputFileContext helmFileCtx) {
-    return isValuesFile(helmFileCtx) || isChartFile(helmFileCtx) || isTplFile(helmFileCtx);
+    return isValuesFile(helmFileCtx) || isTplFile(helmFileCtx);
```
Comment (mstachniuk) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmParser.java:98]: I guess that thanks to this change we will be able now to raise issues in Chart.yaml 🤔 
Please verify that, and if yes, then please add a test case for some rule, e.g. TODO check.

**Review by mstachniuk — APPROVED** [review]
LGTM

File: iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/deprecated_code.yaml (hunk 2 above)
Comment (petertrr) [reply iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/deprecated_code.yaml:29]: A similar ticket is https://sonarsource.atlassian.net/browse/SONARIAC-1309, however now that we can parse the Go AST, we can parse the `.tpl` files as well and execute checks on them. I created a ticket for that: https://sonarsource.atlassian.net/browse/SONARIAC-1616

File: iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/_helpers.tpl (hunk 1 above)
Comment (petertrr) [reply iac-extensions/kubernetes/src/test/resources/checks/DeprecatedCodeCheck/helm/helm3/templates/_helpers.tpl:4]: This function is actually to get the value that will be in the `apiVersion: ...` field of a Deployment. It uses `KubeVersion` to determine a suitable `apiVersion`

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmParser.java (hunk 3 above)
Comment (petertrr) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/plugin/HelmParser.java:98]: Yes, added a TODO comment in ITs

**Pull request merged**